aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -153,6 +153,79 @@ __global__ void compute_grad_weight(
   grad_weight_per_segment[id * stride + startFeature] = weight;
 }
 
+// Fused kernel that combines compute_grad_weight and sum_and_scatter using atomic adds.
+// This eliminates the serialization bottleneck when num_of_segments is small.
+// Each partial segment atomically adds its contribution directly to grad_weight.
+template <typename scalar_t, typename index_t>
+__global__ void compute_grad_weight_fused(
```

Comment:
nit: maybe we could go with a more descriptive kernel name here (imagine seeing this kernel name e.g., in nsys nvprof)